velox/functions/sparksql/specialforms/DecimalRound.cpp

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/DecimalRound.h"
+
+#include "velox/expression/ConstantExpr.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+template <typename R, typename A>
+class DecimalRoundFunction : public exec::VectorFunction {
+ public:
+  explicit DecimalRoundFunction(
+      int32_t scale,
+      uint8_t inputPrecision,
+      uint8_t inputScale,
+      uint8_t resultPrecision,
+      uint8_t resultScale)
+      : scale_(scale),
+        inputPrecision_(inputPrecision),
+        inputScale_(inputScale),
+        resultPrecision_(resultPrecision),
+        resultScale_(resultScale) {
+    auto [p, s] = getResultPrecisionScale(inputPrecision, inputScale, scale);
+    VELOX_USER_CHECK_EQ(p, resultPrecision);
+    VELOX_USER_CHECK_EQ(s, resultScale);
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    applyRoundRows(rows, args, resultType, context, result);
+  }
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+ private:
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    context.ensureWritable(rows, resultType, result);
+    result->clearNulls(rows);
+    return result->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  void applyRoundRows(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const {
+    VELOX_USER_CHECK(
+        args[0]->isConstantEncoding() || args[0]->isFlatEncoding(),
+        "Single-arg deterministic functions receive their only argument as flat or constant vector.");
+    auto rawResults = prepareResults(rows, resultType, context, result);
+    if (args[0]->isConstantEncoding()) {
+      // Fast path for constant vectors.
+      auto constant = args[0]->asUnchecked<ConstantVector<A>>()->valueAt(0);
+      context.applyToSelectedNoThrow(
+          rows, [&](auto row) { rawResults[row] = applyRound(constant); });
```

Comment:
Do we need to call applyRound repeatedly? Can we compute it once?

Reply:
Fixed, thanks.